tests/test_integration.py

```diff
@@ -82,150 +109,166 @@ def _create_vm():
+            hv.run(
+                'virsh destroy {vm}; '
+                'virsh undefine {vm}'
+                .format(vm=self.vm_obj['hostname']),
+                warn_only=True,
+            )
+            hv.run(
+                'umount /dev/xen-data/{vm}; '
+                'lvremove -f /dev/xen-data/{vm}'
+                .format(vm=self.vm_obj['hostname']),
+                warn_only=True,
+            )
 
-    def test_buildvm_auto_find_hypervisor(self):
-        self.vm.server_obj['xen_host'] = None
-        self.vm.server_obj.commit()
+    def get_vm_obj(self):
+        vm_obj = Query(
+            {
+                'servertype': 'vm',
+                'hostname': VM_HOSTNAME,
+            },
+        )
+        return vm_obj.get()
+
+    def check_vm_present(self):
+        # Operate on fresh object
+        vm = VM(self.get_vm_obj())
+
+        for hv in self.hvs:
+            if hv.server_obj['hostname'] == vm.server_obj['xen_host']:
+                # Is it on correct HV?
+                self.assertEqual(hv.vm_defined(vm), True)
+                self.assertEqual(hv.vm_running(vm), True)
+            else:
+                # Is it gone from other HVs after migration?
+                self.assertEqual(hv.vm_defined(vm), False)
+                hv.run('test ! -b /dev/xen-data/{}'.format(vm.fqdn))
+
+        # Is VM itself alive and fine?
+        fqdn = vm.run('hostname -f').strip()
+        self.assertEqual(fqdn, vm.fqdn)
+        self.assertEqual(vm.server_obj.is_dirty(), False)
 
-        buildvm(self.vm.server_obj['hostname'])
+    def check_vm_absent(self, hv_name=None):
+        # Operate on fresh object
+        vm = VM(self.get_vm_obj())
 
-        # Make sure we have updated objects after change.
-        self.vm.reload()
-        self.hv = Hypervisor(self.vm.server_obj['xen_host'])
+        if not hv_name:
+            hv_name = vm.server_obj['xen_host']
 
-        self._check_vm(self.hv, self.vm)
+        for hv in self.hvs:
+            if (hv.server_obj['hostname'] == hv_name):
+                self.assertEqual(hv.vm_defined(vm), False)
+                hv.run('test ! -b /dev/xen-data/{}'.format(vm.fqdn))
 
 
 class BuildTest(IGVMTest):
+    """Test many possible VM building scenarios."""
+
     def setUp(self):
-        self.hv = Hypervisor(HV1)
-        self.vm = _create_vm()
-        _clean_vm([self.hv], self.vm)
+        super(BuildTest, self).setUp()
+        # Normally build tests happen on the 1st HV
+        self.vm_obj['xen_host'] = self.hvs[0].server_obj['hostname']
+        self.vm_obj.commit()
+        self.vm = VM(self.get_vm_obj())
 
     def tearDown(self):
-        _clean_vm([self.hv], self.vm)
-
-    def test_simple(self):
-        buildvm(self.vm.server_obj['hostname'])
-
-        self.assertEqual(self.vm.hypervisor.fqdn, self.hv.fqdn)
-        self._check_vm(self.hv, self.vm)
-
-    def test_simple_stretch(self):
-        self.vm.server_obj.update({
+        super(BuildTest, self).tearDown()
```

Comment:
Same

Reply:
No, this one is not the same. Although I removed methods which *only* call `super()`.

Reply:
True, sorry for the noise.